product/roundhouse/databases/DefaultDatabase.cs

```diff
@@ -238,6 +238,7 @@ public virtual void insert_script_run(string script_name, string sql_to_run, str
             try
             {
                 retry_policy.ExecuteAction(() => repository.save_or_update(script_run));
+                scripts_cache[script_name] = script_run;

```

Comment:
I think that this line alone will provide 99% of the benefits of this P/R.

Reply:
Yeah I agree with that.  The extra hashing optimisation was a small benefit.

Reply:
I'd like to have you test a more surgical version of this.
- Add the line above
- Have get_script_run() just return null (if we start with a full cache and keep it updated, we can trust the cache)
- Rerun your timing test

If that gets you more than 90% of the way there, I'd prefer to take that route.

Reply:
Not sure I follow?

Reply:
Sure I'll proceed as advised.
For me, number 2 has huge benefits because of the way that we use roundhouse and the amount of data actually stored in the roundhouse schema in the actual script column.

Reply:
Also, as an FYI, when I made the changes originally I did them in that order.
And 1 made about 50% of the difference
2 made about 30%
And 3 made about 20%
If I raise these changes (using lazy instead of new model) as 3 prs all extending each other with a screenshot of the difference between current and each pr, you can then made a call as to which of the 3 prs to take.
Does that sound reasonable?
I'm more that happy to put that work in, the net benefits to me and my company on this pr are huge!

Reply:
Yes, that sounds like a reasonable approach. I believe that I can look at the first one on Tuesday or Wednesday.